python/kserve/kserve/storage.py

```diff
@@ -151,8 +151,9 @@ def _download_s3(uri, temp_dir: str):
         #    if awsAnonymousCredential env var true, passed in via config
         # 2. Environment variables
         # 3. ~/.aws/config file
+        endpoint_url = os.getenv("AWS_ENDPOINT_URL") or None
         s3 = boto3.resource('s3',
-                            endpoint_url=os.getenv("AWS_ENDPOINT_URL", "http://s3.amazonaws.com"),
+                            endpoint_url=endpoint_url,
```

Comment:
what happens when `endpoint_url` is set to None here?

Reply:
As I mention in the commit message, boto3 will first try to use the global HTTPS endpoint, that is `https://s3.amazonaws.com` and then it will get redirected to the region of the bucket.

I tried the same snippet with storage-initializer with debug logging enabled:
```
boto3.set_stream_logger('', logging.DEBUG)
```
And without AWS_REGION, AWS_DEFAULT_REGION, AWS_ENDPOINT_URL set.

Indeed the first request was to the global endpoint:
```2022-08-16 06:58:43,456 botocore.endpoint [DEBUG] Making request for OperationModel(name=ListObjects) with params: {'url_path': '/MYBUCKET', 'query_string': {'prefix': 'serving/models/sklearn/1.0/model', 'encoding-type': 'url'}, 'method': 'GET', 'headers': {'User-Agent': 'Boto3/1.18.32 Python/3.7.11 Linux/5.4.181-99.354.amzn2.x86_64 Botocore/1.21.32 Resource'}, 'body': b'', 'url': 'https://s3.amazonaws.com/MYBUCKET?prefix=serving%2Fmodels%2Fsklearn%2F1.0%2Fmodel&encoding-type=url', 'context': {'client_region': 'us-east-1', 'client_config': <botocore.config.Config object at 0x7f46bf4b8290>, 'has_streaming_input': False, 'auth_type': None, 'encoding_type_auto_set': True, 'signing': {'bucket': 'MYBUCKET'}}}
...
2022-08-16 06:58:44,608 botocore.endpoint [DEBUG] Sending http request: <AWSPreparedRequest stream_output=False, method=GET, url=https://MYBUCKET.s3.amazonaws.com/?prefix=serving%2Fmodels%2Fsklearn%2F1.0%2Fmodel&encoding-type=url,...
...
2022-08-16 06:58:43,457 botocore.utils [DEBUG] URI updated to: https://MYBUCKET.s3.amazonaws.com/?prefix=serving%2Fmodels%2Fsklearn%2F1.0%2Fmodel&encoding-type=url
...
2022-08-16 06:58:44,608 urllib3.connectionpool [DEBUG] Starting new HTTPS connection (1): MYBUCKET.s3.amazonaws.com:443
2022-08-16 06:58:44,722 urllib3.connectionpool [DEBUG] https://MYBUCKET.s3.amazonaws.com:443 "GET /?prefix=serving%2Fmodels%2Fsklearn%2F1.0%2Fmodel&encoding-type=url HTTP/1.1" 400 None

```

Then it got redirected to the region of the bucket:
```
2022-08-16 06:58:44,723 botocore.parsers [DEBUG] Response headers: {'x-amz-bucket-region': 'eu-west-1',...
2022-08-16 06:58:44,726 botocore.utils [DEBUG] S3 client configured for region us-east-1 but the bucket MYBUCKET is in region eu-west-1; Please configure the proper region to avoid multiple unnecessary redirects and signing attempts.
...
2022-08-16 06:58:44,726 botocore.utils [DEBUG] URI updated to: https://MYBUCKET.s3.eu-west-1.amazonaws.com/?prefix=serving%2Fmodels%2Fsklearn%2F1.0%2Fmodel&encoding-type=url
...
2022-08-16 06:58:44,968 urllib3.connectionpool [DEBUG] https://MYBUCKET.s3.eu-west-1.amazonaws.com:443 "GET /?prefix=serving%2Fmodels%2Fsklearn%2F1.0%2Fmodel&encoding-type=url HTTP/1.1" 200 None
```

Reply:
@dimara I think the line `endpoint_url = os.getenv("AWS_ENDPOINT_URL") or None` is a bit confusing to read because it seems like we are trying to set the endpoint to `None`. What do you think about this instead?

```
kwargs = {
"service_name": s3,
"config": Storage.get_S3_config()
}

endpoint_url = os.getenv("AWS_ENDPOINT_URL") or None
if endpoint_url:
  kwargs["AWS_ENDPOINT_URL"] = endpoint_url

s3 = boto3.resource(**kwargs)

Reply:
I am not sure whether we need `AWS_ENDPOINT_URL` or not. Because when I tested, the boto3 automatically figured out the endpoint URL. I just did `s3 = boto3.resource('s3')` and it worked. 
I think we don't need to do this `endpoint_url = os.getenv("AWS_ENDPOINT_URL") or None` as `os.getenv` automatically returns None if the env variable does not exist. so only this is enough `endpoint_url = os.getenv("AWS_ENDPOINT_URL")`

Reply:
Afaik, it is needed for users who are using private VPC and use privatelink to access S3 in commercial regions

Reply:
s3 is not just for AWS, it can also be on prem S3 API compatible storage which needs this environment to override with private links.

Reply:
ye I think `endpoint_url = os.getenv("AWS_ENDPOINT_URL")` is good enough.

Reply:
@dimara lets address this comment and merge this PR whenever you get a chance?

Reply:
@surajkota @nishank-lily @yuzisun thanks for the feedback. Let me comment on the above.

I used `os.getenv("AWS_ENDPOINT_URL") or None` to cover the case where the AWS_ENDPOINT_URL is an empty string. I understand this is possible if `endpoint_url` is not set in the storage secret json because of this line:
https://github.com/kserve/kserve/blob/3994c637d3ecedea3c64be969961af3f1fa9a3cc/python/kserve/kserve/storage.py#L115

Passing `endpoint_url=None` to `boto3.resource()` is the default value based on the method's prototype (see https://boto3.amazonaws.com/v1/documentation/api/latest/reference/core/session.html#boto3.session.Session.resource).

I see that you suggest to pass `endpoint_url` kwarg only if it is set (and not empty).  So I guess what you want is:
```
endpoint_url = os.getenv("AWS_ENDPOINT_URL")
if endpoint_url:
  kwargs["endpoint_url"] = endpoint_url
```

So, do you want me to make this change?

Reply:
I think to be consistent we should update line 115 to None as well, so you think it is cleaner this way?

Reply:
@yuzisun line 115 is about setting an environment variable. IMO the equivalent of Python's `None` is either set AWS_ENDPOINT_URL envvar to empty string (as the code currently does) or not set at all. So I wouldn't make any change there. So the question is whether we want to pass `endpoint_url` to boto3 *only* if the AWS_ENDPOINT_URL is set and not empty (see snippet above).

Reply:
@yuzisun @surajkota I updated the PR with the changes you suggested above, that is avoid having the confusing `os.getenv("AWS_ENDPOINT_URL") or None`. Are you OK with them? Can we merge this now?